openmc/tallies.py

```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+        else:
+            all_filters = [self.filters, other.filters]
+            for self_filter, other_filter in itertools.product(*all_filters):
+                new_filter = CrossFilter(self_filter, other_filter, binary_op)
+                new_tally.add_filter(new_filter)
+
+        # Generate score "outer products"
+        if self.scores == other.scores:
+            for self_score in self.scores:
+                new_tally.add_score(self_score)
+        else:
+            all_scores = [self.scores, other.scores]
+            for self_score, other_score in itertools.product(*all_scores):
+                new_score = CrossScore(self_score, other_score, binary_op)
+                new_tally.add_score(new_score)
+
+        # Generate nuclide "outer products"
+        if self.nuclides == other.nuclides:
+            for self_nuclide in self.nuclides:
+                new_tally.nuclides.append(self_nuclide)
+        else:
+            all_nuclides = [self.nuclides, other.nuclides]
+            for self_nuclide, other_nuclide in itertools.product(*all_nuclides):
+                new_nuclide = CrossNuclide(self_nuclide, other_nuclide, binary_op)
+                new_tally.add_nuclide(new_nuclide)
+
+    def _align_tally_data(self, other):
+        """Aligns data from two tallies for tally arithmetic.
+
+        This is a helper method to construct a dict of dicts of the "aligned"
+        data arrays from each tally for tally arithmetic. The method analyzes
+        the filters, scores and nuclides in both tally's and determines how to
+        appropriately align the data for vectorized arithmetic. For example,
+        if the two tallies have different filters, this method will use NumPy
+        'tile' and 'repeat' operations to the new data arrays such that all
+        possible combinations of the data in each tally's bins will be made
+        when the arithmetic operation is applied to the arrays.
+
+        Parameters
+        ----------
+        other : Tally
+            The tally to outer product with this tally
+
+        Returns
+        -------
+        dict
+            A dictionary of dictionaries to "aligned" 'mean' and 'std. dev'
+            NumPy arrays for each tally's data.
+
+        """
+
+        self_mean = copy.deepcopy(self.mean)
+        self_std_dev = copy.deepcopy(self.std_dev)
+        other_mean = copy.deepcopy(other.mean)
+        other_std_dev = copy.deepcopy(other.std_dev)
+
+        if self.filters != other.filters:
+
+            # Determine the number of paired combinations of filter bins
+            # between the two tallies and repeat arrays along filter axes
+            self_num_filter_bins = self.num_filter_bins
+            other_num_filter_bins = other.num_filter_bins
+            num_filter_bins = self_num_filter_bins * other_num_filter_bins
+            self_repeat_factor = num_filter_bins / self_num_filter_bins
+            other_tile_factor = num_filter_bins / other_num_filter_bins
```

Comment:
Haha, yes indeed! This code was a mess at first as I tried different ideas out, and I didn't realize that it had simplified to this point.